File: src/main/java/ru/citeck/ecos/references/PsiElementValueResolver.java
package ru.citeck.ecos.references;

import com.intellij.openapi.extensions.ExtensionPointName;
import com.intellij.psi.PsiElement;

public interface PsiElementValueResolver<T extends PsiElement> {

    ExtensionPointName<PsiElementValueResolver<?>> EP_NAME =
            ExtensionPointName.create("ru.citeck.ecos.psiElementValueResolver");

    Class<T> getPsiElementType();

    Object getValue(T psiElement);

    @SuppressWarnings("unchecked")
    default Object getPsiElementValue(PsiElement psiElement) {
        if (getPsiElementType().isInstance(psiElement)) {
            return getValue((T) psiElement);
        }
        return null;
    }

}

File: src/main/kotlin/ru/citeck/idea/references/PsiElementValueResolver.kt
package ru.citeck.idea.references

import com.intellij.openapi.extensions.ExtensionPointName
import com.intellij.psi.PsiElement

interface PsiElementValueResolver<T : PsiElement> {

    companion object {
        val EP_NAME: ExtensionPointName<PsiElementValueResolver<*>> =
            ExtensionPointName.create("ru.citeck.idea.psiElementValueResolver")
    }

    fun getPsiElementType(): Class<T>

    fun getValue(psiElement: T): Any?

    fun getPsiElementValue(psiElement: PsiElement): Any? {
        if (getPsiElementType().isInstance(psiElement)) {
            @Suppress("UNCHECKED_CAST")
            return getValue(psiElement as T)
        }
        return null
    }
}
